File: android/app/src/main/java/dev/pepotech/pepomote/ui/components/PadCross.kt
package dev.pepotech.pepomote.ui.components

import android.view.HapticFeedbackConstants
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.gestures.awaitEachGesture
import androidx.compose.foundation.gestures.awaitFirstDown
import androidx.compose.foundation.gestures.drag
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.GenericShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.platform.LocalView
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import dev.pepotech.pepomote.control.ButtonState
import dev.pepotech.pepomote.control.DpadModel
import dev.pepotech.pepomote.control.UiSounds
import dev.pepotech.pepomote.ui.theme.PepoColors
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

/**
 * Cruceta de una sola pieza, como la del Mando de Wii: una cruz con las
 * puntas redondeadas y una marca en relieve en cada brazo, sin flechas. El
 * dedo se sigue mientras está apoyado ([DpadModel]): deslizarlo cambia de
 * dirección sin levantarlo, las esquinas entre dos brazos son diagonales y
 * el centro muerto es pequeño. El brazo pulsado se ilumina; cada dirección
 * nueva hace clic (háptica y sonido), como en un mando de verdad. En modo
 * puntero el receptor hace ↑/↓ = flechas del PC y ←/→ = atrás/adelante del
 * navegador. Con [sideways] (mando de lado en un juego) manda los botones
 * del mando girado con el IR a la izquierda ([SidewaysDpad]).
 */
@Composable
fun PadCross(sizeDp: Dp, sideways: Boolean = false) {
    val view = LocalView.current
    var dirs by remember { mutableIntStateOf(0) }
    val glow = PepoColors.Glow
    val mark = PepoColors.CardBorder

    Box(
        modifier = Modifier
            .size(sizeDp)
            .shadow(3.dp, PlusShape, clip = false)
            .background(PepoColors.Card, PlusShape)
            .border(1.5.dp, PepoColors.CardBorder, PlusShape)
            .drawBehind {
                val g = DpadGeometry(size.width / 2f)
                val c = center
                for (dir in 0 until 4) {
                    if (dirs and DpadGeometry.FLAGS[dir] != 0) drawPath(g.arm(c, dir), glow)
                }
                val t = max(2.dp.toPx(), g.a * 0.14f)
                for (dir in 0 until 4) {
                    val (topLeft, markSize) = g.mark(c, dir, t)
                    drawRoundRect(mark, topLeft, markSize, CornerRadius(t / 2f))
                }
            }
            .pointerInput(sideways) {
                var held = 0
                fun release() {
                    for (bit in DpadModel.BUTTON_BITS) if (held and bit != 0) ButtonState.set(bit, false)
                    held = 0
                    dirs = 0
                }
                try {
                    awaitEachGesture {
                        val half = size.width / 2f
                        val c = Offset(half, size.height / 2f)
                        fun apply(pos: Offset) {
                            val d = DpadModel.dirs(pos.x - c.x, pos.y - c.y, half)
                            val bits = DpadModel.buttons(d, sideways)
                            if (bits == held) return
                            for (bit in DpadModel.BUTTON_BITS) {
                                val was = held and bit != 0
                                val now = bits and bit != 0
                                if (was != now) ButtonState.set(bit, now)
                            }
                            // Algo nuevo pulsado (no solo soltado): clic
                            if (bits and held.inv() != 0) {
                                view.performHapticFeedback(HapticFeedbackConstants.KEYBOARD_TAP)
                                UiSounds.blip()
                            }
                            held = bits
                            dirs = d
                        }
                        val down = awaitFirstDown()
                        down.consume()
                        apply(down.position)
                        drag(down.id) { change ->
                            change.consume()
                            apply(change.position)
                        }
                        release()
                    }
                } finally {
                    // Cambio de [sideways] o salida de la pantalla con el dedo puesto
                    release()
                }
            }
    )
}

/**
 * Geometría de la cruz (los mismos números que en iOS y en el móvil Linux):
 * media anchura [half] → brazos de media anchura [a] (un tercio), puntas
 * redondeadas con [rOut] y esquinas interiores con [rIn].
 */
internal class DpadGeometry(val half: Float) {
    val a = half / 3f
    val rOut = a * 0.55f
    val rIn = a * 0.28f

    /** Contorno completo: 12 esquinas en sentido horario desde la interior superior izquierda. */
    fun outline(c: Offset): Path {
        val h = half
        val pts = listOf(
            -a to -a, -a to -h, a to -h, a to -a, h to -a, h to a,
            a to a, a to h, -a to h, -a to a, -h to a, -h to -a
        )
        val radii = FloatArray(12) { if (it % 3 == 0) rIn else rOut }
        return roundedPolygon(pts.map { Offset(c.x + it.first, c.y + it.second) }, radii)
    }

    /** Brazo [dir] (0 ↑, 1 →, 2 ↓, 3 ←): de la arista del cuadrado central a la punta. */
    fun arm(c: Offset, dir: Int): Path {
        val up = listOf(-a to -a, -a to -half, a to -half, a to -a)
        val pts = up.map { (x, y) ->
            var p = x to y
            repeat(dir) { p = -p.second to p.first } // 90 grados en sentido horario
            Offset(c.x + p.first, c.y + p.second)
        }
        return roundedPolygon(pts, floatArrayOf(0f, rOut, rOut, 0f))
    }

    /** Marca en relieve del brazo [dir], a lo largo del brazo y cerca de la punta: esquina y tamaño. */
    fun mark(c: Offset, dir: Int, t: Float): Pair<Offset, Size> {
        val len = a * 0.7f
        val at = half * 0.72f
        return when (dir) {
            0 -> Offset(c.x - t / 2f, c.y - at - len / 2f) to Size(t, len)
            1 -> Offset(c.x + at - len / 2f, c.y - t / 2f) to Size(len, t)
            2 -> Offset(c.x - t / 2f, c.y + at - len / 2f) to Size(t, len)
            else -> Offset(c.x - at - len / 2f, c.y - t / 2f) to Size(len, t)
        }
    }

    companion object {
        /** Dirección de [DpadModel] de cada brazo (0 ↑, 1 →, 2 ↓, 3 ←). */
        val FLAGS = intArrayOf(DpadModel.UP, DpadModel.RIGHT, DpadModel.DOWN, DpadModel.LEFT)
    }
}

/** Polígono con cada esquina redondeada (curva cuadrática; radio ≤ la mitad del lado más corto). */
internal fun roundedPolygon(p: List<Offset>, r: FloatArray): Path {
    val path = Path()
    val n = p.size
    for (i in 0 until n) {
        val prev = p[(i + n - 1) % n]
        val cur = p[i]
        val next = p[(i + 1) % n]
        val rr = min(r[i], min(dist(prev, cur), dist(cur, next)) / 2f)
        val start = towards(cur, prev, rr)
        val end = towards(cur, next, rr)
        if (i == 0) path.moveTo(start.x, start.y) else path.lineTo(start.x, start.y)
        path.quadraticTo(cur.x, cur.y, end.x, end.y)
    }
    path.close()
    return path
}

private fun dist(a: Offset, b: Offset): Float = sqrt((a.x - b.x) * (a.x - b.x) + (a.y - b.y) * (a.y - b.y))

private fun towards(from: Offset, to: Offset, d: Float): Offset {
    val len = dist(from, to)
    return if (len == 0f) from else Offset(from.x + (to.x - from.x) * d / len, from.y + (to.y - from.y) * d / len)
}

/** La cruz como forma de Compose (fondo, borde y sombra). */
internal val PlusShape = GenericShape { size, _ ->
    addPath(DpadGeometry(size.width / 2f).outline(Offset(size.width / 2f, size.height / 2f)))
}

File: ios/PepoMote/UI/Components/PadCross.swift
import SwiftUI

// MARK: - Cruceta

/// Cruceta de una sola pieza, como la del Mando de Wii: una cruz con las
/// puntas redondeadas y una marca en relieve en cada brazo, sin flechas. El
/// dedo se sigue mientras está apoyado (`DpadModel`): deslizarlo cambia de
/// dirección sin levantarlo, las esquinas entre dos brazos son diagonales y
/// el centro muerto es pequeño. El brazo pulsado se ilumina; cada dirección
/// nueva hace clic (háptica y sonido), como en un mando de verdad. En modo
/// puntero el receptor hace ↑/↓ = flechas del PC y ←/→ = atrás/adelante del
/// navegador. Con `sideways` (mando de lado en un juego) manda los botones
/// del mando girado con el IR a la izquierda (`Btn.sideways`).
struct PadCross: View {
    let size: CGFloat
    var sideways = false
    /// Direcciones en pantalla iluminadas (`DpadModel`).
    @State private var dirs = 0
    /// Bits del mando que tenemos pulsados.
    @State private var held: UInt32 = 0

    var body: some View {
        let g = DpadGeometry(half: size / 2)
        let t = max(2, g.a * 0.14)
        ZStack {
            PlusShape()
                .fill(Pepo.card)
                .shadow(color: Color.black.opacity(0.16), radius: 3, x: 0, y: 1.5)
            ForEach(0..<4, id: \.self) { dir in
                if dirs & DpadGeometry.flags[dir] != 0 {
                    ArmShape(dir: dir).fill(Pepo.glow)
                }
            }
            PlusShape().stroke(Pepo.cardBorder, lineWidth: 1.5)
            ForEach(0..<4, id: \.self) { dir in
                let m = g.mark(dir: dir, thickness: t)
                RoundedRectangle(cornerRadius: t / 2, style: .continuous)
                    .fill(Pepo.cardBorder)
                    .frame(width: m.w, height: m.h)
                    .offset(x: m.dx, y: m.dy)
            }
        }
        .frame(width: size, height: size)
        .contentShape(Rectangle())
        .gesture(
            DragGesture(minimumDistance: 0, coordinateSpace: .local)
                .onChanged { v in apply(dx: v.location.x - size / 2, dy: v.location.y - size / 2, half: size / 2) }
                .onEnded { _ in release() }
        )
        .onDisappear { release() }
    }

    private func apply(dx: CGFloat, dy: CGFloat, half: CGFloat) {
        let d = DpadModel.dirs(Float(dx), Float(dy), half: Float(half))
        let bits = DpadModel.buttons(d, sideways: sideways)
        if bits == held { return }
        for bit in DpadModel.buttonBits {
            let was = held & bit != 0
            let now = bits & bit != 0
            if was != now { ButtonState.shared.set(bit, now) }
        }
        // Algo nuevo pulsado (no solo soltado): clic
        if bits & ~held != 0 {
            Haptics.tap()
            UiSounds.shared.blip()
        }
        held = bits
        dirs = d
    }

    private func release() {
        for bit in DpadModel.buttonBits where held & bit != 0 {
            ButtonState.shared.set(bit, false)
        }
        held = 0
        dirs = 0
    }
}

/// Geometría de la cruz (los mismos números que en Android y en el móvil
/// Linux): media anchura `half` → brazos de media anchura `a` (un tercio),
/// puntas redondeadas con `rOut` y esquinas interiores con `rIn`.
struct DpadGeometry {
    let half: CGFloat
    var a: CGFloat { half / 3 }
    var rOut: CGFloat { a * 0.55 }
    var rIn: CGFloat { a * 0.28 }

    /// Dirección de `DpadModel` de cada brazo (0 ↑, 1 →, 2 ↓, 3 ←).
    static let flags = [DpadModel.up, DpadModel.right, DpadModel.down, DpadModel.left]

    /// Contorno completo: 12 esquinas en sentido horario desde la interior superior izquierda.
    func outline(center c: CGPoint) -> Path {
        let h = half
        let pts: [(CGFloat, CGFloat)] = [
            (-a, -a), (-a, -h), (a, -h), (a, -a), (h, -a), (h, a),
            (a, a), (a, h), (-a, h), (-a, a), (-h, a), (-h, -a),
        ]
        let radii = (0..<12).map { $0 % 3 == 0 ? rIn : rOut }
        return roundedPolygon(pts.map { CGPoint(x: c.x + $0.0, y: c.y + $0.1) }, radii)
    }

    /// Brazo `dir` (0 ↑, 1 →, 2 ↓, 3 ←): de la arista del cuadrado central a la punta.
    func arm(center c: CGPoint, dir: Int) -> Path {
        let up: [(CGFloat, CGFloat)] = [(-a, -a), (-a, -half), (a, -half), (a, -a)]
        let pts = up.map { p -> CGPoint in
            var q = p
            for _ in 0..<dir { q = (-q.1, q.0) } // 90 grados en sentido horario
            return CGPoint(x: c.x + q.0, y: c.y + q.1)
        }
        return roundedPolygon(pts, [0, rOut, rOut, 0])
    }

    /// Marca en relieve del brazo `dir`, a lo largo del brazo y cerca de la
    /// punta: centro respecto al de la cruz y tamaño.
    func mark(dir: Int, thickness t: CGFloat) -> (dx: CGFloat, dy: CGFloat, w: CGFloat, h: CGFloat) {
        let len = a * 0.7
        let at = half * 0.72
        switch dir {
        case 0: return (0, -at, t, len)
        case 1: return (at, 0, len, t)
        case 2: return (0, at, t, len)
        default: return (-at, 0, len, t)
        }
    }
}

/// Polígono con cada esquina redondeada (curva cuadrática; radio ≤ la mitad del lado más corto).
func roundedPolygon(_ p: [CGPoint], _ r: [CGFloat]) -> Path {
    var path = Path()
    let n = p.count
    for i in 0..<n {
        let prev = p[(i + n - 1) % n]
        let cur = p[i]
        let next = p[(i + 1) % n]
        let rr = min(r[i], min(polyDist(prev, cur), polyDist(cur, next)) / 2)
        let start = polyTowards(cur, prev, rr)
        let end = polyTowards(cur, next, rr)
        if i == 0 { path.move(to: start) } else { path.addLine(to: start) }
        path.addQuadCurve(to: end, control: cur)
    }
    path.closeSubpath()
    return path
}

private func polyDist(_ a: CGPoint, _ b: CGPoint) -> CGFloat {
    ((a.x - b.x) * (a.x - b.x) + (a.y - b.y) * (a.y - b.y)).squareRoot()
}

private func polyTowards(_ from: CGPoint, _ to: CGPoint, _ d: CGFloat) -> CGPoint {
    let len = polyDist(from, to)
    if len == 0 { return from }
    return CGPoint(x: from.x + (to.x - from.x) * d / len, y: from.y + (to.y - from.y) * d / len)
}

/// La cruz entera (fondo, borde y sombra).
struct PlusShape: Shape {
    func path(in rect: CGRect) -> Path {
        DpadGeometry(half: rect.width / 2).outline(center: CGPoint(x: rect.midX, y: rect.midY))
    }
}

/// Un brazo de la cruz (el pulsado se ilumina).
struct ArmShape: Shape {
    let dir: Int

    func path(in rect: CGRect) -> Path {
        DpadGeometry(half: rect.width / 2).arm(center: CGPoint(x: rect.midX, y: rect.midY), dir: dir)
    }
}
